netket/sampler/autoreg.py

```diff
@@ -0,0 +1,123 @@
+# Copyright 2021 The NetKet Authors - All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from functools import partial
+
+from netket.sampler import Sampler, SamplerState
+from netket.utils import struct
+from netket.utils.types import PRNGKeyT, PyTree
+
+import jax
+from jax import numpy as jnp
+
+
+def vmap_choice(key, a, p, replace=True):
+    """
+    p.shape: (batch, a.shape)
+    Return shape: (batch, )
+    """
+
+    def scan_fun(key, p_i):
+        new_key, key = jax.random.split(key)
+        out_i = jax.random.choice(key, a, replace=replace, p=p_i)
+        return new_key, out_i
+
+    _, out = jax.lax.scan(scan_fun, key, p)
+    return out
+
+
+@struct.dataclass
+class ARDirectSamplerState(SamplerState):
+    σ: jnp.ndarray
+    """current batch of (maybe partially sampled) configurations."""
+    cache: PyTree
+    """auxiliary states, e.g., used to implement fast autoregressive sampling."""
+    key: PRNGKeyT
+    """state of the random number generator."""
+
+    def __repr__(self):
+        return f"{type(self).__name__}(rng state={self.key})"
+
+
+@struct.dataclass
+class ARDirectSampler(Sampler):
```

Comment:
Is it necessary to specify ```Direct``` here in the name? What would be the non-direct AR alternatives?

Reply:
Maybe we can use the potentially saved space and call it `AutoregressiveSampler`, which is easier to understand compared to `AR`?

Reply:
This arose from a discussion with @wdphy16 .
Since there can also be a MonteCarlo Chain sampler based on the autoregressive ansatz, that will be a different sampler so I thought we should differentiate.

Reply:
Uhm true, but then I am not sure ```Direct``` is a good name. I think most people call this ancestral sampling https://ermongroup.github.io/cs228-notes/inference/sampling/ (or also in the Deep Learning book)

Reply:
Yeah, fine, because it's not really direct sampling.
Either is fine by me. It's just that we should not change it in the future.

Reply:
yes the only issue is that ancestral sampling is more general, it should work for arbitrary dependency graph... uhm I don't know, I think that the the other sampling method is just ```nk.sampler.AutoregressiveCluster```and this one is pure ```nk.sampler.Autoregressive```

Reply:
did we get to a conclusion?